design/kanister-kopia-integration.md

```diff
@@ -0,0 +1,594 @@
+2. A Kanister Profile of type `kopia` requires a
+   [Kopia Repository Server](https://kopia.io/docs/repository-server/) running
+   in the same namespace as the Kanister controller
+3. A Repository Server requires a
+   [Kopia Repository](https://kopia.io/docs/repositories/) to be initialized on
+   a backend storage target
+
+Kanister currently lacks documentation and automation to use these features.
+
+## Introducing Kopia
+
+Kopia is a powerful, cross-platform tool for managing encrypted backups in the
+cloud.
+It provides fast and secure backups, using compression, data deduplication, and
+client-side end-to-end encryption.
+It supports a variety of backup storage targets, including object stores, which
+allows users to choose the storage provider that better addresses their needs.
+In Kopia, these storage locations are called repositories.
+It is a lock-free system that allows concurrent multi-client operations
+including garbage collection.
+
+To explore other features of Kopia, see its
+[documentation](https://kopia.io/docs/features/).
+
+## Scope
+
+1. Design and automate the lifecycle of the required Kopia Repository Server.
+2. Add new versions of Kanister Data Functions like BackupData,
+   RestoreData, etc. with Kopia as the primary data mover tool.
+
+## User Experience
+
+- All the new features mentioned in this document will be opt-in only. Existing
+   users will not see any changes in the Kanister controller's behavior.
+- Users will be able to continue using their current Blueprints, switch to the
+   v2 version of the example Blueprints, or use Blueprints with the new version
+   of the Kanister Data Functions.
+- Users opting to use the v2 Blueprints and Blueprints with Kopia-based
+   Kanister Data Functions will be required to follow instructions to set up
+   the required Kopia repository and the repository server before executing the
+   actions.
+- After setting up the repository and the repository server, users can follow
+   the normal workflow to execute actions from the v2 Blueprints. To use the
+   new versions of the Kanister Data Functions, users must specify the version
+   of the function via the ActionSet Action's `preferredVersion` field. This
+   field in an Action is applied to all the Phases in it. If a particular
+   Kanister function is not registered with this version, Kanister will fall
+   back to the default version of the function.
+
+## Detailed Design
+
+### Kanister Data Functions
+
+- Kanister allows mutliple versions of Functions to be registered with the
+   controller.
+- Existing Functions are registered with the default `v0.0.0` version. Find
+   more information
+   [here](https://docs.kanister.io/functions.html#existing-functions).
+- The following Data Functions will be registered with a second
+   version `v1.0.0-alpha`:
+
+   1. BackupData
+   2. BackupDataAll
+   3. BackupDataStats
+   4. CopyVolumeData
+   5. DeleteData
+   6. DeleteDataAll
+   7. RestoreData
+   8. RestoreDataAll
```

Comment:
As discussed, instead of adding versions, we would add new functions with the names having`*WithKopia` suffix.

Reply:
@shlokc9 can you add a note about this is yet to decide or TBD.

Reply:
Sure, I'll add a note about this. 👍🏼 
Maybe a separate design would be needed for handling kopia clients inside Kanister Functions....